gematria/datasets/basic_block_utils.cc

```diff
@@ -42,50 +47,75 @@ unsigned getSuperRegister(unsigned OriginalRegister,
     return OriginalRegister;
 }
 
-std::vector<unsigned> BasicBlockUtils::getUsedRegisters(
-    const std::vector<DisassembledInstruction> &Instructions,
+std::vector<unsigned> getUsedRegisters(
+    const ArrayRef<DisassembledInstruction> Instructions,
     const MCRegisterInfo &RegisterInfo, const MCInstrInfo &InstructionInfo) {
-  std::map<unsigned, bool> UsedInputRegisters;
-  std::map<unsigned, bool> DefinedInBBRegisters;
+  std::set<unsigned> UsedInputRegisters;
+  std::set<unsigned> DefinedInBBRegisters;
   for (const gematria::DisassembledInstruction &Instruction : Instructions) {
     unsigned InstructionDefs =
         InstructionInfo.get(Instruction.mc_inst.getOpcode()).getNumDefs();
     for (unsigned OperandIndex = InstructionDefs;
          OperandIndex < Instruction.mc_inst.getNumOperands(); ++OperandIndex) {
-      if (Instruction.mc_inst.getOperand(OperandIndex).isReg()) {
-        unsigned RegisterNumber =
-            Instruction.mc_inst.getOperand(OperandIndex).getReg();
+      const MCOperand &CurrentOperand =
+          Instruction.mc_inst.getOperand(OperandIndex);
+      if (CurrentOperand.isReg()) {
+        unsigned RegisterNumber = CurrentOperand.getReg();
         if (RegisterNumber == 0) continue;
         unsigned SuperRegister = getSuperRegister(RegisterNumber, RegisterInfo);
         // If the register was already defined within the basic block, we don't
         // need to define it ourselves.
         if (DefinedInBBRegisters.count(SuperRegister) > 0) continue;
-        UsedInputRegisters[SuperRegister] = true;
+        UsedInputRegisters.insert(SuperRegister);
       }
     }
     // We also need to handle instructions that have implict uses.
     for (unsigned ImplicitlyUsedRegister :
          InstructionInfo.get(Instruction.mc_inst.getOpcode()).implicit_uses()) {
-      UsedInputRegisters[getSuperRegister(ImplicitlyUsedRegister,
-                                          RegisterInfo)] = true;
+      UsedInputRegisters.insert(
+          getSuperRegister(ImplicitlyUsedRegister, RegisterInfo));
+    }
+    // Handle instructions that have implicit defs
+    for (unsigned ImplicitlyDefinedRegister :
+         InstructionInfo.get(Instruction.mc_inst.getOpcode()).implicit_defs()) {
+      // Do not define the super register of 8-bit or 16-bit registers as
+      // already being defined, as moving values into those registers does not
+      // zero the rest of the bits in the register, unlike when writing to
+      // 32-bit registers.
+      if (RegisterInfo.getRegClass(X86::GR8RegClassID)
+              .contains(ImplicitlyDefinedRegister) ||
+          RegisterInfo.getRegClass(X86::GR16RegClassID)
+              .contains(ImplicitlyDefinedRegister))
+        continue;
+      DefinedInBBRegisters.insert(
+          getSuperRegister(ImplicitlyDefinedRegister, RegisterInfo));
```

Comment:
Nitpicky: this is repeated almost verbatim below, let's put it e.g. to a lambda and then just call the lambda.